text/3880-deref_into.md

```diff
@@ -0,0 +1,249 @@
+- Feature Name: `deref_into`
+- Start Date: 2025-11-10
+- RFC PR: [rust-lang/rfcs#3880](https://github.com/rust-lang/rfcs/pull/3880)
+- Rust Issue: [rust-lang/rust#3880](https://github.com/rust-lang/rust/issues/3880)
+
+# Summary
+[summary]: #summary
+
+Introduce supertraits `DerefInto` and `DerefMutInto` for `Deref` and `DerefMut` that return their targets by value with independent mutable and immutable targets.
+
+This enables using deref coercion with RAII guards (`RefCell`, `Mutex`, `RwLock`...).
+
+# Motivation
+[motivation]: #motivation
+
+Rust's `Deref` and `DerefMut` traits are special because of the "Deref coercion". [Deref coercion is incredibly important for ergonomics. It's so that we can effectively interact with all of the different types of smart pointers as though they were regular ol' references e.g. allowing Box<T> where &T is expected.](https://www.reddit.com/r/rust/comments/1654y5l/comment/jyc9l4x).
+
+However, the standard `Deref`/`DerefMut` traits are limited in that their output types are fixed references (`&Self::Target` / `&mut Self::Target`).
+
+This makes them incompatible with interior mutability or synchronization types such as `RefCell`, `Mutex`, or `RwLock`, which return RAII guards by value (`Ref` / `RefMut` for `RefCell`, `MutexGuard` for `Mutex`, and `RwLockReadGuard` / `RwLockWriteGuard` for `RwLock`).
+
+
+```rust
+pub struct Foo
+{
+    pub bar: i32,
+}
+
+let cell: RefCell<Foo> = RefCell::new(Foo{ bar: 42 });
+
+let _bar = cell.borrow().bar; // Current way to write it
+
+let _bar = &cell.bar; // Error: Impossible to express right now,
+// but more ergonomic to write
+```
```

Comment:
I don't find this code more clear or ergonomic. Sure, okay, "ergonomic to write" maybe? But while it may be easier to write, it is much harder to read. We should not aspire to be a write-once language.